### PR TITLE
support reference_time for duckling

### PR DESCRIPTION
**Pull request by vinvinod**
**Proposed changes**:
Duckling module also supports passing a context using reference_time argument. We can pass a string in RFC3339 format (Eg: '2008-09-08T22:47:31-07:00').

Example: Parsing "tomorrow" with reference_time set to `2008-09-08T20:30:00-07:00` gives back 
`2008-09-09T00:00:00.000-07:00`

Since Rasa supports duckling in the pipeline, I think this will be a useful addition.

**Status**:
- [ ] ready for code review
- [ ] there are tests for the functionality
- [ ] documentation updated
- [ ] changelog updated

### CODE REVIEW

**Comment by vinvinod** [comment]
The emulator tests are failing. I shall make the corrections.
But I noticed thats there is an issue with duckling. It doesnt handle timezone offsets with non-zero minutes (https://github.com/wit-ai/duckling_old/issues/209)
So will this feature be useful?

File: _pytest/test_components.py [hunk 1]
```diff
@@ -85,7 +85,7 @@ def test_all_arguments_can_be_satisfied_during_parse(component_class, default_co
 
     # All available context arguments that will ever be generated during parse
     component = component_builder.create_component(component_class.name, default_config)
-    context_arguments = {"text": None}
+    context_arguments = {"text": None, "ref_time": ""}
```
Comment (tmbo) [inline _pytest/test_components.py]: Let's use `None` as the default

Comment (tmbo) [inline _pytest/test_components.py]: Also, I'd like to rename this to something that expresses that this is the time of the message. Do you have a suggestion here, maybe just `time`?

File: rasa_nlu/data_router.py [hunk 2]
```diff
@@ -175,7 +175,7 @@ def parse(self, data):
                 raise InvalidModelError("No model found with alias '{}'. Error: {}".format(alias, e))
 
         model = self.model_store[alias]
-        response = model.parse(data['text'])
+        response = model.parse(data['text'], data['ref_time'])
```
Comment (tmbo) [inline rasa_nlu/data_router.py]: Should be optional - eg. `data.get('time')`

File: rasa_nlu/extractors/duckling_extractor.py [hunk 3]
```diff
@@ -77,12 +78,28 @@ def pipeline_init(self, language):
             except ValueError as e:     # pragma: no cover
                 raise Exception("Duckling error. {}".format(e))
 
-    def process(self, text, entities):
+    def process(self, text, entities, ref_time=""):
```
Comment (tmbo) [inline rasa_nlu/extractors/duckling_extractor.py]: `None` instead of `""`

File: rasa_nlu/extractors/duckling_extractor.py [hunk 4]
```diff
@@ -77,12 +78,28 @@ def pipeline_init(self, language):
...
         # type: (Text, List[Dict[Text, Any]]) -> Dict[Text, Any]
 
         extracted = []
         if self.duckling is not None:
-            matches = self.duckling.parse(text)
+            # check if ref_time is a datetime is in a known format
+            # If its not parseable, make it empty
+            if not ref_time:
+                ref_time = ""
```
Comment (tmbo) [inline rasa_nlu/extractors/duckling_extractor.py]: How about setting it to the current server time?

File: _pytest/test_components.py (hunk 1 above)
Comment (vinvinod) [inline _pytest/test_components.py]: `time` should be fine

File: rasa_nlu/extractors/duckling_extractor.py (hunk 4 above)
Comment (vinvinod) [inline rasa_nlu/extractors/duckling_extractor.py]: Duckling takes the current system time and timezone by default.

Comment (tmbo) [inline rasa_nlu/extractors/duckling_extractor.py]: Right, but we might use that time for other things in the future (e.g. for logging)

**Comment by vinvinod** [comment]
The usefulness of this feature is that the duckling can parse the input string by using the desired context (time with timezone) and we get the exact timestamp as the parsed output. I think we should pass the timezone info somehow for parsing. If we just pass the timestamp in milliseconds, the zone info is lost. 

We can also take timezone info separately! 
`time` for time in milliseconds and `timezone` for taking the offset in minutes like the below request.
`http://localhost:5000/parse?text=Call%20me%20tomorrow%20morning%20at%208&time=1496055692128&timezone=-330`
Then we also have to validate the timezone offset value.

**Comment by vinvinod** [comment]
Another way to standardise would be to take just `time` in millis and always assume `timezone` as UTC. So that the users can convert the parsed time to whatever zone they want in their applications.

**Comment by tmbo** [comment]
Alright, lets use a UTC ISO time string then. :+1:

**Comment by tmbo** [comment]
Sorry @vinvinod I totally forgot about this one. If you don't mind fixing the merge conflicts with master, I could merge it afterwards.

File: rasa_nlu/extractors/duckling_extractor.py [hunk 5]
```diff
@@ -78,12 +80,34 @@ def cache_key(cls, model_metadata):
 
         return cls.name + "-" + model_metadata.language
 
-    def process(self, text, entities):
+    def pipeline_init(self, language):
```
Comment (tmbo) [inline rasa_nlu/extractors/duckling_extractor.py]: This function needs to be removed (I don't think it is necessary anyways), it most likely is an artifact from resolving the merge conflicts with master. This function is also the reason the tests fail.

**Comment by tmbo** [comment]
nice work 😃 :+1:

**Comment by vinvinod** [comment]
👍

**Pull request merged**